// @ecash/lib/dist/hash.js
"use strict";
// Copyright (c) 2024 The Bitcoin developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.
Object.defineProperty(exports, "__esModule", { value: true });
exports.__setHashes = exports.sha512Hasher = exports.sha256Hasher = exports.sha512 = exports.shaRmd160 = exports.sha256d = exports.sha256 = void 0;
const HASHES = {};
function sha256(data) {
    return HASHES.sha256(data);
}
exports.sha256 = sha256;
function sha256d(data) {
    return HASHES.sha256d(data);
}
exports.sha256d = sha256d;
function shaRmd160(data) {
    return HASHES.shaRmd160(data);
}
exports.shaRmd160 = shaRmd160;
function sha512(data) {
    return HASHES.sha512(data);
}
exports.sha512 = sha512;
function sha256Hasher() {
    return new HASHES.Sha256H();
}
exports.sha256Hasher = sha256Hasher;
function sha512Hasher() {
    return new HASHES.Sha512H();
}
exports.sha512Hasher = sha512Hasher;
function __setHashes(hashes) {
    HASHES.sha256 = hashes.sha256;
    HASHES.sha256d = hashes.sha256d;
    HASHES.shaRmd160 = hashes.shaRmd160;
    HASHES.sha512 = hashes.sha512;
    HASHES.Sha256H = hashes.Sha256H;
    HASHES.Sha512H = hashes.Sha512H;
}
exports.__setHashes = __setHashes;
